### PR TITLE
Feature/certificates

### PR DESCRIPTION
**Pull request by dhelmert**
### Reason for this PR
- Hyperledger registration and enrollment is needed for proper blockchain usage

### Changes in this PR
- Added CertificateService to handle registration and enrollment
- Added UserService communication with Certificate

- [x] Test written
- [x] Changelog updated

### CODE REVIEW

**Comment by dhelmert** [comment]
A quick look in the test coverage report for the CertificateServiceImpl reveals that we are missing all negative tests, i.e. "NotFound", "OwnerMismatch" and "NotEnrolled". Tests for these probably need to be added before this branch is merged

File: product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        if (validationErrors.nonEmpty) {
+          throw new UC4Exception(422, DetailedError(ErrorType.Validation, validationErrors))
+        }
+
+        getCertificateUser(username).flatMap {
+          case (Some(enrollmentId), Some(enrollmentSecret), _, _) =>
+            val certificate = enrollmentManager.enrollSecure(caURL, tlsCert, enrollmentId, enrollmentSecret, pmcsrRaw.certificateSigningRequest, adminUsername, walletPath, channel, chaincode, networkDescriptionPath)
+            entityRef(username).ask[Confirmation](replyTo => SetCertificateAndKey(certificate, pmcsrRaw.encryptedPrivateKey, replyTo)).map {
+              case Accepted => (ResponseHeader(202, MessageProtocol.empty, List()), Done)
+              case _        => throw UC4Exception.InternalServerError
```
Comment (david-buderus) [inline product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala:70]: You should forward the message of the rejection

File: product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    //REGISTER
+    "register a CertificateUser" in {
+      val probe = createTestProbe[Confirmation]()
+      val ref = spawn(CertificateBehaviour.create(PersistenceId("fake-type-hint", "fake-id-5")))
+      ref ! RegisterUser("enrollmentId", "enrollmentSecret", probe.ref)
+      probe.expectMessage(Accepted)
+    }
+
+    //SETCERTANDKEY
```
Comment (david-buderus) [inline product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala:72]:
```suggestion
    //set certificate and key
```

File: product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      ref ! SetCertificateAndKey("certificate", EncryptedPrivateKey("key", "iv", "salt"), probe2.ref)
+      probe2.expectMessage(Accepted)
+
+      //fetch a registered user
+      val probe3 = createTestProbe[(Option[String], Option[String], Option[String], Option[EncryptedPrivateKey])]()
+      ref ! GetCertificateUser(probe3.ref)
+      probe3.expectMessage((Some("enrollmentId"), Some("enrollmentSecret"), Some("certificate"), Some(EncryptedPrivateKey("key", "iv", "salt"))))
+    }
+
+    //REGISTER
```
Comment (david-buderus) [inline product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala:64]:
```suggestion
    //register user
```

**Review by david-buderus — CHANGES_REQUESTED** [review]

File: product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala (hunk 2 above)
Comment (dhelmert) [reply product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala:72]: In other tests we used upper case comments to denote the categories of the tests, such as "GET", "ADD", "DELETE TESTS". While we maybe should change the comments to names more conform to previous naming schemes, e.g. "GET" "REGISTER" and "SET", we shouldn't divert from the current test category naming scheme

Comment (david-buderus) [reply product_code/certificate_service/impl/src/test/scala/de/upb/cs/uc4/certificate/impl/CertificateStateSpec.scala:72]:
```suggestion
    //SET CERTIFICATE AND KEY
```
Than like this?

File: product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        if (validationErrors.nonEmpty) {
+          throw new UC4Exception(422, DetailedError(ErrorType.Validation, validationErrors))
+        }
+
+        getCertificateUser(username).flatMap {
+          case (Some(enrollmentId), Some(enrollmentSecret), None, None) =>
+            val certificate = enrollmentManager.enrollSecure(caURL, tlsCert, enrollmentId, enrollmentSecret, pmcsrRaw.certificateSigningRequest, adminUsername, walletPath, channel, chaincode, networkDescriptionPath)
+            entityRef(username).ask[Confirmation](replyTo => SetCertificateAndKey(certificate, pmcsrRaw.encryptedPrivateKey, replyTo)).map {
+              case Accepted     => (ResponseHeader(202, MessageProtocol.empty, List()), Done)
+              case e: Exception => throw UC4Exception.InternalServerError("Failed to set certificate", e.getMessage)
```
Comment (david-buderus) [inline product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala:70]: This is false.
You will never get an exception here.
You will get a RejectedWithReason.

**Review by david-buderus — CHANGES_REQUESTED** [review]

File: product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala (hunk 4 above)
Comment (dhelmert) [reply product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala]: Good point, this is now resolved

**Review by ldklenner — APPROVED** [review]

**Review by david-buderus — APPROVED** [review]

**Review by RHelmert — APPROVED** [review]

**Pull request merged**